src/crimson/osd/osd_operations/pg_advance_map.cc

```diff
@@ -89,32 +89,145 @@ seastar::future<> PGAdvanceMap::start()
+  DEBUG(" pg_num change in e{} {} -> {}", next_map->get_epoch(),
+                 old_pg_num, new_pg_num);
+  std::set<spg_t> children;
+  if (new_pg_num && new_pg_num > old_pg_num) {
+    if (pg->pgid.is_split(
+	    old_pg_num,
+	    new_pg_num,
+	    &children)) {
+      co_await split_pg(children, next_map);
+    }
+  }
+  co_return;
+}
+
+
+seastar::future<> PGAdvanceMap::split_pg(
+    std::set<spg_t> split_children,
+    cached_map_t next_map)
+{
+  LOG_PREFIX(PGAdvanceMap::split_pg);
+  DEBUG("{}: start", *this);
+  auto pg_epoch = next_map->get_epoch();
+  DEBUG("{}: epoch: {}", *this, pg_epoch);
+
+  co_await seastar::coroutine::parallel_for_each(split_children, [this, &next_map,
+  pg_epoch, FNAME] (auto child_pgid) -> seastar::future<> {
+    children_pgids.insert(child_pgid);
+
+    // Map each child pg ID to a core
+    auto core = co_await shard_services.create_split_pg_mapping(child_pgid, seastar::this_shard_id());
+    DEBUG(" PG {} mapped to {}", child_pgid.pgid, core);
+    DEBUG(" {} map epoch: {}", child_pgid.pgid, pg_epoch);
+    auto map = next_map;
+    auto child_pg = co_await shard_services.make_pg(std::move(map), child_pgid, true);
+
+    DEBUG(" Parent pgid: {}", pg->get_pgid());
+    DEBUG(" Child pgid: {}", child_pg->get_pgid());
+    unsigned new_pg_num = next_map->get_pg_num(pg->get_pgid().pool());
+    // Depending on the new_pg_num the parent PG's collection is split.
+    // The child PG will be initiated with this split collection.
+    unsigned split_bits = child_pg->get_pgid().get_split_bits(new_pg_num);
+    DEBUG(" pg num is {}, m_seed is {}, split bits is {}",
+	new_pg_num, child_pg->get_pgid().ps(), split_bits);
+
+    co_await pg->split_colls(child_pg->get_pgid(), split_bits, child_pg->get_pgid().ps(),
+                             &child_pg->get_pgpool().info, rctx.transaction);
+    DEBUG(" {} split collection done", child_pg->get_pgid());
+    // Update the child PG's info from the parent PG
+    pg->split_into(child_pg->get_pgid().pgid, child_pg, split_bits);
+
+    co_await handle_split_pg_creation(child_pg, next_map);
+  });
+
+  split_stats(split_pgs, children_pgids);
+  co_return;
+}
+
+seastar::future<> PGAdvanceMap::handle_split_pg_creation(
+    Ref<PG> child_pg,
+    cached_map_t next_map)
+{
+  LOG_PREFIX(PGAdvanceMap::handle_split_pg_creation);
+  // We must create a new Trigger instance for each pg.
+  // The BlockingEvent object which tracks whether a pg creation is complete
+  // or still blocking, cannot be used across multiple pgs.
+  using EventT = PGMap::PGCreationBlockingEvent;
+  using TriggerT = EventT::Trigger<PGAdvanceMap>;
+  EventT event;
+  TriggerT trigger(event, *this);
+  auto fut = shard_services.create_split_pg(std::move(trigger), child_pg->get_pgid());
+  co_await shard_services.start_operation<PGAdvanceMap>(
+      child_pg, shard_services, next_map->get_epoch(),
+      std::move(rctx), true).second;
+  co_await fut.safe_then([this, FNAME] (Ref<PG> child_pgref) {
```

Comment:
Why not just co_await std::move(fut);?  Using safe_then just means introducing another lambda.  Also, why not just co_await create_split_pg?  This construction allows the PGAdvanceMap on child_pg to occur concurrently with create_split_pg.  That seems unnecessary complicated.

Reply:
Thanks for catching that. I'll make the change

Reply:
@athanatos I looked into your suggestion, and the reason I chose this approach is due to a dependency on `PGAdvanceMap`.  
`create_split_pg` waits for the PG to be created, and the associated promise (from `PGCreationState`) is fulfilled in `pg_created`  
([link](https://github.com/ceph/ceph/blob/main/src/crimson/osd/pg_map.cc#L219)), which is called as part of `PGAdvanceMap`.
A similar pattern exists in `ShardServices::get_or_create_pg`, but in that case, we discard the future instead of awaiting it. 

I have tried to explain this in the comments now.